tools/autograd/templates/python_torch_functions.cpp

```diff
@@ -130,12 +132,14 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
           .dtype(scalarType)
           .device(r.device(6))
           .layout(r.layout(5).layout)
-          .requires_grad(r.toBool(7));
+          .requires_grad(r.toBool(8))
+          .pinned_memory(r.toBool(7));
       return wrap(dispatch_arange(start, end, step, options));
     } else {
+      AT_ASSERTM(!r.toBool(7), " `pin_memory` and `out` parameters are incompatible");
```

Comment:
Use of `ASSERT` here is inappropriate. Asserts are for error conditions that occur when we have a bug in our code. A user can easily trigger this condition; thus, an assert should not be used.

Reply:
Is there any particular reason this error check is done, as opposed to the actual implementation of `arange`? Error checks in the Python binding code don't show up in C++ API, and this seems like a reasonable thing to check for C++ too.

Reply:
We also call check_out_type_matches here. And it does exactly same, pre-validation of passed parameters and throwing AT_ERROR. 

There is no need to implement this check inside of cpp function as in case of output tensor it doesn't take options.

`wrap(dispatch_arange(r.scalar(0), r.scalar(1), r.scalar(2), r.tensor(3)).set_requires_grad(r.toBool(8)))`

where 
 0 - start
 1 - end
 2 - step 
 3 - output
 8 - requires_grad

Reply:
OK, if there's no C++ out + TensorOptions variant, that's sufficient for me.

Reply:
> We also call check_out_type_matches here. And it does exactly same, pre-validation of passed parameters and throwing AT_ERROR.

Just double-checking; this comment is in reference to "Use of ASSERT here is inappropriate. Asserts are for error conditions that occur when we have a bug in our code. A user can easily trigger this condition; thus, an assert should not be used." is that right?

Reply:
I guess it wasn't, because this is still broken:

```
>>> torch.arange(2, out=x, pin_memory=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: !r.toBool(5) ASSERT FAILED at ../torch/csrc/autograd/generated/python_torch_functions.cpp:119, please report a bug to PyTorch.  `pin_memory` and `out` parameters are incompatible
```

Reply:
What I'm trying to say is that 

```
arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)
arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)
```
Is incorrect approach (if we are trying to avoid AT_ERROR), with or without `pin_memory` introduced.

The proper approach is to have 4 signatures:

```
arange(Scalar end, *, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)
arange(Scalar start, Scalar end, Scalar step=1, *, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)

arange(Scalar end, *, Tensor out, bool requires_grad=False)
arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out, bool requires_grad=False)
```

In this case `check_out_type_matches` will go away as well as `AT_ASSERTM(!r.toBool(5),`. But this requires bigger review of all function and out of scope.

Reply:
Erm, rewinding a sec: my original ask was that you use `AT_CHECK` rather than `AT_ASSERT`. This is why I posted a snippet of a simple user code snippet triggering the assert. **The assert literally tells the user to file a bug** and we don't want that in this situation. I'm not asking for any restructuring of the code in question. In particular, I'm not asking you to completely remove the error check.

I agree your four signatures is better, and I also agree we shouldn't do it in this PR.